server/jetty/src/main/java/io/deephaven/server/jetty/jsplugin/JsPlugins.java

```diff
@@ -0,0 +1,112 @@
+ *
+ * @see #servletHolder(String)
+ */
+public final class JsPlugins implements JsPluginRegistration {
+
+    private static final String PLUGINS = "plugins";
+    private static final String NAME = "name";
+    private static final String VERSION = "version";
+    private static final String MAIN = "main";
+    private static final String MANIFEST_JSON = "manifest.json";
+
+    /**
+     * Creates a new js plugins instance with a temporary zip filesystem.
+     *
+     * @return the js plugins
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPlugins create() throws IOException {
+        final Path tempDir = Files.createTempDirectory(JsPlugins.class.getName());
+        tempDir.toFile().deleteOnExit();
+        final Path fsZip = tempDir.resolve("fs.zip");
+        fsZip.toFile().deleteOnExit();
+        // Note, the URI needs explicitly be parseable as a directory URL ending in "!/", a requirement of the jetty
+        // resource creation implementation, see
+        // org.eclipse.jetty.util.resource.Resource.newResource(java.lang.String, boolean)
+        final URI uri = URI.create(String.format("jar:file:%s!/", fsZip));
+        final JsPlugins jsPlugins = new JsPlugins(uri);
+        jsPlugins.init();
+        return jsPlugins;
+    }
+
+    private final URI filesystem;
+    private final List<Map<String, String>> plugins;
+
+    private JsPlugins(URI filesystem) {
+        this.filesystem = Objects.requireNonNull(filesystem);
+        this.plugins = new ArrayList<>();
+    }
+
+    /**
+     * Creates a {@link DefaultServlet} servlet holder.
+     *
+     * @param name the name
+     * @return the servlet holder
+     */
+    public ServletHolder servletHolder(String name) {
+        final ServletHolder jsPlugins = new ServletHolder(name, DefaultServlet.class);
+        jsPlugins.setInitParameter("resourceBase", filesystem.toString());
+        jsPlugins.setInitParameter("pathInfoOnly", "true");
+        jsPlugins.setInitParameter("dirAllowed", "false");
+        // etag support picked up via io.deephaven.server.jetty.JettyBackedGrpcServer
+        // jsPlugins.setInitParameter("etags", "true");
+        jsPlugins.setAsyncSupported(true);
+        return jsPlugins;
+    }
+
+    @Override
+    public synchronized void register(JsPlugin jsPlugin) throws IOException {
+        for (Map<String, String> plugin : plugins) {
+            if (jsPlugin.name().equals(plugin.get(NAME))) {
+                throw new IllegalArgumentException(
+                        String.format("js plugin with name '%s' already exists", jsPlugin.name()));
+            }
+        }
+        try (final FileSystem fs = FileSystems.newFileSystem(filesystem, Map.of())) {
+            final Path pluginPath = fs.getPath("/", jsPlugin.name());
+            jsPlugin.copyTo(pluginPath);
+            plugins.add(Map.of(NAME, jsPlugin.name(), VERSION, jsPlugin.version(), MAIN, jsPlugin.main()));
+            writeManifest(fs);
```

Comment:
to confirm, there is no clean way to _not_ call this N+1 times? once when the instance is created (but only via the factory ctor), and once for each plugin added?

Reply:
If we're just concerned about re-writing the manifest multiple times, we can probably come up with a solution to only write it "once".

More generally though, it's not as easy as writing a visitor that collects all the plugins and then installs them in one shot. The lifecycle concerns around `io.deephaven.plugin.type.JsPlugin#copyTo` for clean python interop has python opening and closing the distribution directory for consumption `py/server/deephaven/server/plugin/register.py`:

```python
elif isinstance(plugin, JsPlugin):
    with plugin.distribution_path() as distribution_path:
        self._callback.registerJsPlugin(str(distribution_path), plugin.name, plugin.version, plugin.main)
```

Our avenues for better batch installing either involve 

1. Adding batch registration interfaces in both python and java, plumbing as appropriate
2. Having an extra layer of indirection during installation, then batching up results (for example, copying everything to jimfs, and then batch to zipfs).

I'll try and document these concerns a bit in code, and see if there is an easy way to solve for just the manifest.

Reply:
I did take a stab at trying to be more efficient wrt the manifest; but it wasn't very elegant and I figured it would be better to try and solve it in a more complete way outside of this PR.

https://github.com/deephaven/deephaven-core/issues/3007